common-lib/src/main/java/com/google/cloud/tools/intellij/analytics/UsageTracker.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.analytics;
+
+import com.google.cloud.metrics.Event;
+import com.google.cloud.metrics.MetricsException;
+import com.google.cloud.metrics.MetricsSender;
+import com.google.cloud.tools.intellij.flags.PropertiesFileFlagReader;
+import com.google.common.collect.ImmutableMap;
+
+import com.intellij.ide.util.PropertiesComponent;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.updateSettings.impl.UpdateChecker;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Map;
+
+public class UsageTracker {
```

Comment:
the class performs IO, so think of it as inheriting the state of the remote
data it operates in. To test a class that calls into these static methods
would mean unavoidably doing IO which is bad for a unit test.

I'll send you a link off this PR with some guidelines.

On Wed, May 25, 2016 at 12:10 PM, Chanseok Oh <notifications@github.com>
wrote:

> In
> common-lib/src/main/java/com/google/cloud/tools/intellij/analytics/UsageTracker.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/699#discussion_r64600571>
> :
>
> > +import com.google.cloud.metrics.Event;
> > +import com.google.cloud.metrics.MetricsException;
> > +import com.google.cloud.metrics.MetricsSender;
> > +import com.google.cloud.tools.intellij.flags.PropertiesFileFlagReader;
> > +import com.google.common.collect.ImmutableMap;
> > +
> > +import com.intellij.ide.util.PropertiesComponent;
> > +import com.intellij.openapi.diagnostic.Logger;
> > +import com.intellij.openapi.updateSettings.impl.UpdateChecker;
> > +
> > +import org.jetbrains.annotations.NotNull;
> > +import org.jetbrains.annotations.Nullable;
> > +
> > +import java.util.Map;
> > +
> > +public class UsageTracker {
>
> Alright, I actually kind of expected this.
>
> So, I considered this class fundamentally stateless (everything is static),
> but you're suggesting that it's more suitable to to make a single instance
> and register it as a service, right? That makes sense. I'll go ahead and
> change it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/699/files/f5ffe78841982c60aa40e45e6025e1bd55664b1c#r64600571>
>